### PR TITLE
Additional functionality and fixes.

### PR DESCRIPTION
**Pull request by jmaxson-ms**
Additional data retrieval methods and extension methods.
Engine may use multiple plugin paths.
Additional engine creation info.
Fixes XPerf symbols deadlock.
Updated and new tests.

### CODE REVIEW

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/SourceDataCookerDependencyProvider.cs [hunk 1]
```diff
@@ -33,9 +33,22 @@ public TOut QueryOutput<TOut>(DataOutputPath dataOutputPath)
             return (TOut)QueryOutput(dataOutputPath);
         }
 
+        public bool TryQueryOutput<TOut>(DataOutputPath dataOutputPath, out TOut result)
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/SourceDataCookerDependencyProvider.cs:36]: We seem to be using just 'T' on the other methods instead of 'TOut.' Should we use just 'T' here for consistency?

File: src/Microsoft.Performance.SDK.Runtime/TypeExtensions.cs [hunk 2]
```diff
@@ -19,5 +19,17 @@ internal static bool TryGetEmptyPublicConstructor(this Type type, out Constructo
             constructor = null;
             return false;
         }
+
+        internal static bool TryGetEmptyConstructor(this Type type, out ConstructorInfo constructor)
+        {
+            constructor = type.GetConstructor(Type.EmptyTypes);
+            if (constructor != null)
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.SDK.Runtime/TypeExtensions.cs:26]: we can just return constructor != null instead of the if / else, as if the constructor is null, we set it to null and return false anyway.

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        /// <param name="result">
+        ///     The uniquely identified data.
+        /// </param>
+        /// <exception cref="InvalidOperationException">
+        ///     The specified <paramref name="dataOutputPath"/> was not found.
+        /// </exception>
+        /// <exception cref="InvalidCastException">
+        ///     The value at <paramref name="dataOutputPath"/> cannot be cast to <typeparamref name="T"/>.
+        /// </exception>
+        public static void QueryOutput<T>(
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:36]: Why does this extension exist? It returns void and uses an out parameter. Do we need it?

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        ///     The uniquely identified data.
+        /// </param>
+        /// <exception cref="InvalidOperationException">
+        ///     The specified <paramref name="outputId"/> was not found on the data cooker 
+        ///     <paramref name="cookerPath"/>.
+        /// </exception>
+        /// <exception cref="InvalidCastException">
+        ///     The value at <paramref name="identifier"/> cannot be cast to <typeparamref name="T"/>.
+        /// </exception>
+        public static void QueryOutput<T>(
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:160]: Same as for ````void QueryOutput<T>```` at the top of the file.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 5]
```diff
@@ -80,5 +102,23 @@ public string ExtensionDirectory
...
         /// </summary>
         public VersionChecker Versioning { get; internal set; }
+
+        /// <summary>
+        ///     Gets or sets the name of the runtime on which the application is built.
+        /// </summary>
+        /// <remarks>
+        ///     Defauls to "Microsoft.Performance.Toolkit.Engine".
+        /// </remarks>
+        public string RuntimeName { get; set; } = "Microsoft.Performance.Toolkit.Engine";
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:112]: Do we want to add any validation against null / whitespace here?

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 6]
```diff
@@ -14,53 +17,72 @@ namespace Microsoft.Performance.Toolkit.Engine
...
                 {
-                    throw new InvalidExtensionDirectoryException(value);
+                    throw new InvalidExtensionDirectoryException(directory);
                 }
 
-                this.extensionDirectory = dirInfo.FullName;
+                directories.Add(dirInfo.FullName);
             }
+
+            this.ExtensionDirectories = directories.AsReadOnly();
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:77]: Does it make sense to dedupe this list here?

**Review by zachnew-msft — CHANGES_REQUESTED** [review]

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/SourceDataCookerDependencyProvider.cs (hunk 1 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/SourceDataCookerDependencyProvider.cs:36]: This class is already using T as a type parameter, so for these methods, we use TOut instead.

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs (hunk 3 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:36]: It makes it so that you don't need to specify the type T when making the call. For example, this works...

QueryOutput(path, out bool result);

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs (hunk 4 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:160]: Same answer.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs (hunk 6 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:77]: Zack and I spoke offline. This makes sense, but DirectoryInfo doesn't implement IEquatable or similar and making this work across platforms would be challenging. We're just going to leave this for now.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs (hunk 5 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:112]: The engine handles this. I think null/empty should be ok here.

                string runtimeName = !string.IsNullOrWhiteSpace(createInfo.RuntimeName)
                    ? createInfo.RuntimeName
                    : "Microsoft.Performance.Toolkit.Engine";

**Review by zachnew-msft — DISMISSED** [review]

**Review by zachnew-msft — APPROVED** [review]

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 7]
```diff
@@ -80,5 +102,23 @@ public string ExtensionDirectory
...
+
+        /// <summary>
+        ///     Gets or sets the application name.
+        /// </summary>
+        public string ApplicationName { get; set; }
+
+        /// <summary>
+        ///     Gets or sets a means of logging information.
+        /// </summary>
+        public ILogger Logger { get; set; }
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:122]: I think it would make more sense to have this be a logger factory, i.e. Func<Type, ILogger> so that the user is in control of whether the same logger gets reused across all data sources or not. As this appears to be implemented in the engine, if this property is set, then it is used for all Custom Data Sources, but if it is not set, then a new Logger is created for each Custom Data Source. Using a factory function provides that same flexibility but still allows the caller to do something else if desiered.

Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs]: Updated.

File: src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/CompositeDataCookerReferenceTests.cs [hunk 8]
```diff
@@ -118,7 +118,20 @@ public object QueryOutput(DataOutputPath identifier)
             {
                 throw new NotImplementedException();
             }
+
+            public bool TryQueryOutput<T>(DataOutputPath identifier, out T result)
+            {
+                result = default;
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/CompositeDataCookerReferenceTests.cs:124]: is there a reason these don't just also `throw new NotImplementedException`

File: src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/TestClasses/TestSourceDataCooker.cs [hunk 9]
```diff
@@ -33,12 +33,42 @@ public TestSourceDataCooker(TestSourceDataCookerContext context)
 
         public T QueryOutput<T>(DataOutputPath identifier)
         {
-            throw new System.NotImplementedException();
+            if (Reflector == null)
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/TestClasses/TestSourceDataCooker.cs:36]: Should we have a separate text fixture for the reflection tests?

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/DataCookers/CompositeDataCookerReference.cs [hunk 10]
```diff
@@ -92,7 +91,7 @@ private CompositeDataCookerReference(Type type)
 
             reference = null;
 
-            if (candidateType.IsPublicAndInstantiatableOfType(typeof(ICompositeDataCookerDescriptor)))
+            if (candidateType.IsInstantiatableOfType(typeof(ICompositeDataCookerDescriptor)))
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/DataCookers/CompositeDataCookerReference.cs:94]: what's the reason for this change?

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs [hunk 11]
```diff
@@ -64,6 +60,34 @@ public object QueryOutput(DataOutputPath dataOutputPath)
             }
         }
 
+        /// <inheritdoc/>
+        public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T result)
+        {
+            if (!string.IsNullOrWhiteSpace(dataOutputPath.SourceParserId))
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:66]: we should probably get the dataoutput path PR in and then check the DataOutputPathType instead

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs [hunk 12]
```diff
@@ -102,6 +126,11 @@ private void ValidateSourceDataCooker(DataOutputPath dataOutputPath)
             }
         }
 
+        private bool TryValidateSourceDataCooker(DataOutputPath dataOutputPath)
+        {
+            return this.extensionDependencies.RequiredSourceDataCookerPaths.Contains(dataOutputPath.CookerPath);
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:131]: do we want to avoid using `.CookerPath`?

**Review by zachnew-msft — APPROVED** [review]

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/DataCookers/CompositeDataCookerReference.cs (hunk 10 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/DataCookers/CompositeDataCookerReference.cs:94]: It's required to support non-public custom data source, which is something we want specifically for DataLayer, but would also be good to have in general.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs (hunk 12 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:131]: I don't know, is there?

File: src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/CompositeDataCookerReferenceTests.cs (hunk 8 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/CompositeDataCookerReferenceTests.cs:124]: I'm just in the habit of not throwing from Try methods :). I'll add it.

File: src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/TestClasses/TestSourceDataCooker.cs (hunk 9 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime.Tests/Extensibility/TestClasses/TestSourceDataCooker.cs:36]: Perhaps, but I don't want to update for this PR as this was already being used.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs (hunk 11 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:66]: Whichever is checked in first, the other should update this. Either way, I don't care.

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+            DataCookerPath cookerPath,
+            string outputId,
+            out object result)
+        {
+            return retrieval.TryQueryOutput(new DataOutputPath(cookerPath, outputId), out result);
+        }
+
+        /// <summary>
+        ///     Retrieves data by name, typecast to the given type.
+        ///     The caller is coupled to the data extension and expected to know the
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:101]: what exactly does "The caller is coupled to the data extension" mean? Is this just saying to call this generic method the caller needs to know at compile time what T is? Isn't that implied with a typeparam in the signature?

File: src/Microsoft.Performance.SDK/Processing/CustomDataProcessorBaseWithSourceParser.cs [hunk 14]
```diff
@@ -131,41 +131,103 @@ public void EnableCooker(ISourceDataCookerFactory sourceDataCookerFactory)
...
-        /// <param name="dataOutputPath">Path to the data output</param>
-        /// <returns>Data output of the specified type</returns>
+        /// <inheritdoc/>
+        /// <remarks>
+        ///     Returns data from a source data cooker registered with <see cref="SourceProcessingSession"/>.
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        ///     <paramref name="dataOutputPath"/> does not target the data processor.
+        /// </exception>
         public TOutput QueryOutput<TOutput>(DataOutputPath dataOutputPath)
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK/Processing/CustomDataProcessorBaseWithSourceParser.cs:141]: why do we even have methods for querying data cookers on a custom data processor? seems like we're blurring boundaries between object functionality

File: src/Microsoft.Performance.Toolkit.Engine/Engine.cs [hunk 15]
```diff
@@ -805,6 +808,22 @@ protected void ThrowIfProcessed()
...
+        /// </summary>
+        /// <param name="type">
+        ///     The <c>Type</c> for which the logger is created.
+        /// </param>
+        /// <returns>
+        ///     An instance of <see cref="ILogger"/>.
+        /// </returns>
+        protected ILogger CreateLogger(Type type)
+        {
+            Debug.Assert(!this.isDisposed);
```
Comment (mslukebo) [inline src/Microsoft.Performance.Toolkit.Engine/Engine.cs:822]: is this line redundant with the line below?

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 16]
```diff
@@ -14,53 +17,72 @@ namespace Microsoft.Performance.Toolkit.Engine
...
+        /// <param name="extensionDirectory">
+        ///     The extension directory from which the runtime instance is to load plugins.
+        /// </param>
+        /// <remarks>
+        ///     If no extension directory is specified, the current directory is used.
+        /// </remarks>
+        /// <exception cref="InvalidExtensionDirectoryException">
+        ///     Thrown if <paramref name="extensionDirectory"/> is an invalid directory path.
+        /// </exception>
+        public EngineCreateInfo(string extensionDirectory = null)
```
Comment (mslukebo) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:32]: thoughts on a separate parameterless ctor instead of using a default?

File: src/Microsoft.Performance.Toolkit.Engine/Engine.cs (hunk 15 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/Engine.cs:822]: No, this can make it easier to find while debugging.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 17]
```diff
@@ -80,5 +102,26 @@ public string ExtensionDirectory
...
         /// </summary>
         public VersionChecker Versioning { get; internal set; }
+
+        /// <summary>
+        ///     Gets or sets the name of the runtime on which the application is built.
+        /// </summary>
+        /// <remarks>
+        ///     Defaults to "Microsoft.Performance.Toolkit.Engine".
+        /// </remarks>
+        public string RuntimeName { get; set; } = "Microsoft.Performance.Toolkit.Engine";
```
Comment (mslukebo) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:112]: are we able to grab the assembly name dynamically/at compile time instead of hard-coding this?

File: src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs [hunk 18]
```diff
@@ -19,6 +18,7 @@ namespace Microsoft.Performance.Toolkit.Engine
     ///     Represents the results of processing data using the <see cref="Engine"/>.
     /// </summary>
     public sealed class RuntimeExecutionResults
+        : ICookedDataRetrieval
```
Comment (mslukebo) [inline src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:21]: was this class already implementing this interface implicitly? I don't see anything added to the class.

Why are "results" able to retrieve data? I feel like that should be happening directly through the engine/runtime, not through its final result. but that might be a bigger change

File: src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs (hunk 13 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Extensibility/ICookedDataRetrievalExtensions.cs:101]: It means that it must know the return type - there can be no compile time type check. And if that return type changes, the caller must be updated.

File: src/Microsoft.Performance.SDK/Processing/CustomDataProcessorBaseWithSourceParser.cs (hunk 14 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/CustomDataProcessorBaseWithSourceParser.cs:141]: This does blur things a bit, you're correct and we can chat offline. But I'm going to resolve this comment and it's not new functionality to this PR.

File: src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs (hunk 18 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:21]: This class was already implementing the interface, yes. Adding the interface allows the interface extensions to work with this class.

The results are how clients of the engine retrieve any data from the output. @zachnew-msft designed it this way, and it seems to work well. But if you have some thoughts on how to improve it, happy to chat.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs (hunk 12 above)
Comment (mslukebo) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:131]: I was just thinking if we're trying to use the rep of a cooker path less maybe we should be passing around/using the actual DataOutPath objects instead. We'd could just implement IComparable/IEquatable on the paths

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs (hunk 16 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:32]: sure, i'm ok with that.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs (hunk 17 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:112]: That's probably a good idea.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs (hunk 12 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:131]: This isn't a string, it's a DataCookerPath, which does implement IEquatable<>. And We want to compare the cooker paths, not the DataOutputPath, which is why we're using that property.

Comment (mslukebo) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/FilteredDataRetrieval.cs:131]: Ah, my bad. I thought this DataCookerPath.CookerPath which was the string.

File: src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs [hunk 19]
```diff
@@ -14,53 +17,87 @@ namespace Microsoft.Performance.Toolkit.Engine
...
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="EngineCreateInfo"/> class.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">
+        ///     Thrown is <paramref name="extensionDirectories"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="InvalidExtensionDirectoryException">
+        ///     Thrown if <paramref name="extensionDirectories"/> contains an invalid directory path.
+        /// </exception>
+        public EngineCreateInfo(IEnumerable<string> extensionDirectories)
```
Comment (mslukebo) [inline src/Microsoft.Performance.Toolkit.Engine/EngineCreateInfo.cs:61]: missing a <param> in docstring 😃

**Review by mslukebo — DISMISSED** [review]

**Review by zachnew-msft — APPROVED** [review]

**Review by mslukebo — APPROVED** [review]

**Review by zachnew-msft — APPROVED** [review]
